### PR TITLE
Fix stacking invalid order requests

### LINKED ISSUE #2663: Errors in checkout stack into multiple orders per submit
## Current behavior
The more times you submit an order with invalid data (i.e. only numbers in the name fields) the more times there will be a submit order request sent to the backend. First submit will send only one request, another submit w/o fixing errors will send the order twice, etc.

## Expected behavior
Order requests having error are being repeated over and over.

## Steps to reproduce the issue
1. try to place an order with invalid user data (i.e. a few digits in name fields)
2. hit submit button several times
3. the more times you hit the button the more requests will be sent to the backend incrementally.

# Can you handle fixing this bug by yourself?

- [x] YES but not anytime soon, feel free to fix it
- [ ] NO

# Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer.
Pick one option.

- [ ] This is a bug report for test version on https://test.storefrontcloud.io - In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
- [ ] This is a bug report for current Release Candidate version on https://next.storefrontcloud.io - In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
- [x] This is a bug report for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version hotfix - In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

## Environment details

- Browser: Chrome 72
- OS: Ubuntu 18.04
- Node: 8
- Code Version: develop

### PR DESCRIPTION
**Pull request by lukeromanowicz**
### Related issues
#2663 #2740 #2751

closes #2663 #2740 

### Short description and why it's useful
It prevents repeating order requests by disabling order caching in case of invalid request body (http code 400). 

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [x] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/doc/Upgrade%20notes.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)
- [x] I read the [TypeScript Action Plan](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/TypeScript%20Action%20Plan.md) and adjusted my PR according to it
- [x] I read about [Vue Storefront Modules](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/api-modules/about-modules.md) and I am aware that every new feature should be a module

### CODE REVIEW

**Comment by pkarw** [comment]
@lukeromanowicz for sure You will add it - but we probably need to check the response statuses and not stack the orders with failing validation, but probably stack the orders with server error - and surely with the lack of internet connection. However, when the server is down because of traffic peak in 9/10 cases it will return 502/503 error - so kind of 5xxx

**Comment by lukeromanowicz** [comment]
This PR fixes only handling HTTP 400 responses to order create requests with an invalid address on directOrderSync turned on, so a proper popup is being shown and bad requests are not stacking in the queue. In my opinion, it's not possible to make it work well with directOrderSync being off without #2753

**Comment by pkarw** [comment]
In the rare case of the offline mode order error we should show the notification like: “There was an error processing You’re offline order. Probably the ordered products went out of the stock. Please re-order the products or contact us”

In this case we have two options for a work around:
A) show this notification, remove order from the queue, add the products to shopping cart and redirect to cart (Remake order exactly as it works ok User Account - order history - remake order)
B) show this notification, remove order from the queue

Of course A is preferred - I bet You can make it :-)

Then some day we’ll implement the #2753 

We can do B to 1.9 and A to 1.10rc1 if you feel more safe with that but we
Probably shouldn’t leave this issue unaddressed as we’re aware of it in the end...

File: core/modules/order/store/actions.ts [hunk 1]
```diff
@@ -45,26 +47,48 @@ const actions: ActionTree<OrderState, RootState> = {
...
         Vue.prototype.$bus.$emit('notification-progress-stop')
-        if (task.resultCode !== 500) {
+
+        if (task.resultCode === 200) {
           order.transmited = true
           commit(types.ORDER_PLACE_ORDER, order) // archive this order but not trasmit it second time
           commit(types.ORDER_LAST_ORDER_WITH_CONFIRMATION, { order: order, confirmation: task.result })
           Vue.prototype.$bus.$emit('order-after-placed', { order: order, confirmation: task.result })
+
+          return task
```
Comment (pkarw) [inline core/modules/order/store/actions.ts]: Aren’t we showing which fields do containa errors  as Youu initially planned?

**Comment by pkarw** [comment]
We can merge it in but handling the error messages in case of validation issue plus offline order errors (A or B) is crucial.

File: core/modules/order/store/actions.ts (hunk 1 above)
Comment (lukeromanowicz) [reply core/modules/order/store/actions.ts]: @pkarw no, it will be added in separate PR

Comment (pkarw) [reply core/modules/order/store/actions.ts]: Ok

**Comment by pkarw** [comment]
OK could you move the comments out of this PR to kind of issue to not loose the ideas then? @lukeromanowicz

**Comment by lukeromanowicz** [comment]
@pkarw I've implemented B) for now

File: core/lib/sync/task.ts [hunk 2]
```diff
@@ -118,7 +118,7 @@ function _internalExecute (resolve, reject, task: Task, currentToken, currentCar
             Vue.prototype.$bus.$emit('modal-show', 'modal-signup')
           }
         }
-        if (!task.silent && (jsonResponse.result && jsonResponse.result.code !== 'ENOTFOUND' && !silentMode)) {
+        if (!task.silent && (typeof jsonResponse.result === 'string' && jsonResponse.result.code !== 'ENOTFOUND' && !silentMode)) {
```
Comment (patzick) [inline core/lib/sync/task.ts]: `(typeof jsonResponse.result === 'string' && jsonResponse.result.code !== 'ENOTFOUND'`
if it is a string, then it doesn't have code

Comment (lukeromanowicz) [reply core/lib/sync/task.ts]: you're right. I've changed that

**Pull request merged**